content/blog/2023/sqlalchemy-for-django-developers.md

```diff
@@ -0,0 +1,335 @@
+        back_populates="user", cascade="all, delete-orphan"
+    )
+
+class Address(DeclarativeBase):
+    __tablename__ = "address"
+
+    id: Mapped[int] = mapped_column(primary_key=True)
+    email_address: Mapped[str]
+    user_id: Mapped[int] = mapped_column(ForeignKey("user_account.id"))
+
+    user: Mapped["User"] = relationship(back_populates="addresses")
+```
+
+Notice you need to define both the `user_id` column and the `user` relationship
+in the `Address` model, something that Django does automatically. Additionally,
+you also need to define the "other side" of the relationship as
+`User.addresses`.
+
+Once your models are in place you can query the `Address` model and join it with
+the `User` model:
+
+```python
+from sqlalchemy import select
+
+stmt = (
+    select(Address)
+    .join(Address.user)
+    .where(User.name == "sandy")
+)
+sandy_addresses = session.scalars(stmt).all()
+```
+
+There is no double underscore syntax like in Django, and no "related managers"
+that allow you to get related objects from a given model instance. Instead we
+need to use a `join` to bring in the `User` model and then filter on the `name`
+attribute.
+
+One clear advantage of SQLAlchemy is that you can instantiate model instances
+and relate them before they are saved, and then save them all at once:
+
+```python
+spongebob = User(
+    name="spongebob",
+    fullname="Spongebob Squarepants",
+    addresses=[Address(email_address="spongebob@sqlalchemy.org")],
+)
+sandy = User(
+    name="sandy",
+    fullname="Sandy Cheeks",
+    addresses=[
+        Address(email_address="sandy@sqlalchemy.org"),
+        Address(email_address="sandy@squirrelpower.org"),
+    ],
+)
+patrick = User(name="patrick", fullname="Patrick Star")
+session.add_all([spongebob, sandy, patrick])
+session.commit()
+```
+
+In Django you would need to save all the users first, then save the addresses
+with the `user` attribute set to the corresponding user. Here we can nest them
+naturally and SQLAlchemy will take care of the rest.
+
+## Migrations are not Built-in
+
+Django comes with a built-in migration system that allows you to create and
+apply migrations with a single command. This is definitely one of the most
+important features of Django and I was surprised to find out that SQLAlchemy
+does not have a built-in migration system. Instead, they maintain a separate
+package called [Alembic] to manage migrations.
+
+[Alembic]: https://alembic.sqlalchemy.org/en/latest/
+
+We won't go into details here, but the basic substitutions are:
+
+- `./manage.py makemigrations` becomes `alembic revision --autogenerate`
+- `./manage.py migrate` becomes `alembic upgrade head`
+- `./manage.py migrate app <migration number>` becomes `alembic upgrade
+  <migration hash>` if going forward, or `alembic downgrade <migration
+  hash>` if going back
+
+To date I don't know how to manage migrations in databases that have models
+defined by third party packages.
```

Comment:
Would it be worth giving an example of when you might run into this?

Reply:
Added in ed2dcbfb674fb0820bcd55a53c9b73038f330438